src/Admin/hooks/useAsyncSelectOption.ts

```diff
@@ -0,0 +1,133 @@
+
+/**
+ * Custom hook for handling async option selection with pagination and search
+ *
+ * @unreleased
+ */
+export function useAsyncSelectOptions({
+    recordId,
+    entity,
+    selectedOptionRecord,
+    endpoint,
+    recordsFormatter = (records: any) => records,
+    optionFormatter,
+    queryParams,
+    perPage = 30,
+    resetOnChange = false,
+}: AsyncSelectOptionsConfig): UseAsyncSelectOptionReturn {
+    const [page, setPage] = useState(0);
+    const [selectedOption, setSelectedOption] = useState<Option | null>(null);
+    const [error, setError] = useState<Error | null>(null);
+
+    // Reset page when reset property changes
+    useEffect(() => {
+        if (resetOnChange !== undefined) {
+            setPage(0);
+        }
+    }, [resetOnChange]);
+
+    useEffect(() => {
+        if (selectedOptionRecord && recordId) {
+            setSelectedOption(optionFormatter(selectedOptionRecord));
+        } else if (!recordId) {
+            setSelectedOption(null);
+        }
+    }, [selectedOptionRecord, recordId]);
+
+    // Load options function for AsyncPaginate
+    const loadOptions = useCallback(async (searchInput: string) => {
+        const currentPage = searchInput ? 1 : page + 1;
+
+        const params = new URLSearchParams({
+            ...queryParams,
+            per_page: perPage.toString(),
+            page: currentPage.toString(),
+            ...(searchInput && {search: searchInput}),
+        });
+
+        setError(null);
+
+        try {
+            const records = recordsFormatter(await apiFetch<[]>({
+                path: `${endpoint}?${params.toString()}`,
+            }));
+
+            const newOptions = (records || []).map(optionFormatter);
+
+            // Update page state
+            if (searchInput !== '') {
+                setPage(1);
+            } else if (!searchInput) {
+                setPage(currentPage);
+            }
+
+            const hasMoreResults = (records?.length || 0) >= perPage;
+
+            return {
+                options: newOptions,
+                hasMore: hasMoreResults,
+            };
+        } catch (err) {
+            const loadError = err instanceof Error ? err : new Error(`Failed to load ${entity}`);
+            setError(loadError);
+            console.error(`Failed to load ${entity}`, loadError);
+
+            return {
+                options: [],
+                hasMore: false,
+            };
+        }
+    }, [page]);
+
+    // Map options for menu (deduplication and ordering)
+    const mapOptionsForMenu = useCallback(
+        (options: Option[]) => filterOptionsForSelect(options, selectedOption),
+        [selectedOption],
+    );
+
+    return {
+        selectedOption,
+        loadOptions,
+        mapOptionsForMenu,
+        error,
+    };
+}
+
+export type Option = {
+    value: number;
+    label: string;
+}
+
+export type AsyncSelectOptionsConfig = {
+    recordId: number | null;
+    entity: string;
```

Comment:
Is there any reason why we should keep this? Right now is only used in an error message, but we can make that message more generic.